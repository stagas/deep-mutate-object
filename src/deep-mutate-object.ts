export const deepMutate = (obj: any, walkFn: (key: string, value: any, holder: any) => [string, any]): any =>
  Object.entries(obj).reduce((p, [k, v]) => {
    const [key, value] = walkFn(k, v, obj)

    p[key] = typeof value === 'object' && value !== null
      ? deepMutate(value, walkFn)
      : value

    return p
  }, (Array.isArray(obj) ? [] : {}) as any)
