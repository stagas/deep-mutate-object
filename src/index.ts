export * from './deep-mutate-object'
