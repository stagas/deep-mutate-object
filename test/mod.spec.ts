import { deepMutate } from '../src/deep-mutate-object'

describe('deepMutate(obj, walkFn)', () => {
  it('simple', () => {
    const result = deepMutate({
      foo: 'bar',
    }, (key, value) => {
      return [key + 'zoo', value + 'land']
    })
    expect(result).toMatchSnapshot()
  })

  it('deep', () => {
    const result = deepMutate({
      foo: 'bar',
      some: {
        other: 'deeper',
      },
    }, (key, value) => {
      return [key + 'zoo', typeof value === 'string' ? value + 'land' : value]
    })
    expect(result).toMatchSnapshot()
  })

  it('array', () => {
    const result = deepMutate({
      foo: 'bar',
      some: {
        other: [1, 2, {
          inside: 'array',
        }],
      },
    }, (key, value, holder) => {
      return [Array.isArray(holder) ? key : key + 'zoo', typeof value === 'string' ? value + 'land' : value]
    })
    expect(result).toMatchSnapshot()
  })
})
