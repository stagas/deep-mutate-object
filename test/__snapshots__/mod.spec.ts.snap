
exports[`deepMutate(obj, walkFn) simple 1`] = `
Object {
  "foozoo": "barland",
}
`;

exports[`deepMutate(obj, walkFn) deep 1`] = `
Object {
  "foozoo": "barland",
  "somezoo": Object {
    "otherzoo": "deeperland",
  },
}
`;

exports[`deepMutate(obj, walkFn) array 1`] = `
Object {
  "foozoo": "barland",
  "somezoo": Object {
    "otherzoo": Array [
      1,
      2,
      Object {
        "insidezoo": "arrayland",
      },
    ],
  },
}
`;
